feat: add telex integration for enhanced logging

johnafariogun opened the pull request:
---

## Description  

This PR integrates Telex logging into the global exception handler, ensuring that all unexpected errors are logged and reported to Telex in real time. It enhances observability and debugging capabilities by capturing error details like timestamp, request method, request path, status code, and error message.  

## Related Issue (Link to issue ticket)  

closes [#1024](https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/1024)

## Motivation and Context  

This change is required to improve error tracking and debugging by automatically reporting critical exceptions to Telex. By using asynchronous logging with `httpx.AsyncClient`, this ensures that application performance is not affected by logging operations.  

## How Has This Been Tested?  

- Added unit tests to verify that errors are logged and sent to Telex.  
- Simulated various exceptions to confirm that logs include the correct details.  
- Ensured that performance remains unaffected during logging operations.  

![telex_logger](https://github.com/user-attachments/assets/ec89cab1-b493-41df-bcf1-342b24bfd246)


## Types of changes  

- [ ] Bug fix (non-breaking change which fixes an issue)  
- [x] New feature (non-breaking change which adds functionality)  
- [ ] Breaking change (fix or feature that would cause existing functionality to change)  

## Checklist:  

- [x] My code follows the code style of this project.  
- [x] My change requires a change to the documentation.  
- [x] I have updated the documentation accordingly.  
- [x] I have read the **CONTRIBUTING** document.  
- [x] I have added tests to cover my changes.  
- [x] All new and existing tests passed.  

---

johnafariogun committed:
chore: save local changes before pulling from remote repo

johnafariogun committed:
Merge branch 'dev' of github.com:johnafariogun/hng_boilerplate_python_fastapi_web into feat/telex-integration-for-enhanced-logging
chore: remote repo ahead of local repo, need merge in order to push

johnafariogun committed:
feat: integrate Telex logging for global exception handling

johnafariogun committed:
Merge branch 'hngprojects:dev' into feat/telex-integration-for-enhanced-logging

johnafariogun committed:
chore: added telex webhook url to .env.sample

johnafariogun committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging

johnafariogun committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging

johnafariogun commented:
Thanks for the review 🙏🏾

samuelogboye commented on api/utils/send_logs.py:
```diff
@@ -0,0 +1,34 @@
+                "status_code": 500,
+                "error_message": f"An unexpected error occurred: {exc}",
+            }
+        ),
```
I think converting a dictionary to string will result in a less readable format. 
why not consider either using a proper nested JSON structure or formatting a clean string message

samuelogboye commented on api/utils/send_logs.py:
```diff
@@ -0,0 +1,34 @@
+
+    async with httpx.AsyncClient() as client:
+        try:
+            await client.post(TELEX_WEBHOOK_URL, json=error_data)
```
There's no check if TELEX_WEBHOOK_URL is empty or None, which could cause directionless errors

samuelogboye replied:
You don't check the response status from Telex, or do you assume it will always be successful?
you might want to log if an error occurs

samuelogboye commented on api/utils/send_logs.py:
```diff
@@ -0,0 +1,34 @@
+                "event_name": "server_error",
+                "request_method": request_method,
+                "request_path": request_url_path,
+                "status_code": 500,
```
This hardcodes status_code as 500, which may not be accurate for all errors.
extract the code if available

johnafariogun commented:
I'll get right to that sir

johnafariogun committed:
fix: extract status code dynamically and handle missing webhook URL

- Extracts `status_code` dynamically from the exception, defaulting to 500 if unavailable.
- Adds a safeguard to check if `TELEX_WEBHOOK_URL` is set before sending logs.
- Improves error logging for better observability.

johnafariogun committed:
test: fix assertion errors and ensure correct mocking

johnafariogun committed:
test: added mock webhook properly

johnafariogun committed:
refactor(tests): replace with-patch syntax with decorator for mocking

johnafariogun committed:
refactor(tests): removed redundant comments

johnafariogun committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging

johnafariogun committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging

johnafariogun committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging

johnafariogun committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging

johnafariogun committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging

johnafariogun committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging

johnafariogun committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging

joboy-dev committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging

johnafariogun committed:
chore: added telex channel webhook

johnafariogun committed:
chore: removed telex channel webhook

joboy-dev committed:
Merge branch 'dev' into feat/telex-integration-for-enhanced-logging
